### DISCUSSION TITLE
"Serial exception" on startup

### DISCUSSION THREAD

**Discussion by jstarek** [g4ixt/QtTinySA#26] [General]
I am trying to run the software on an Ubuntu 22.04 LTS system and get the following message on startup (with an already switched-on TinySA connected):

```
$ python QtTinySA.py 
QSqlDatabase: QSQLITE driver not loaded
QSqlDatabase: available drivers: 
Database open: False
QSqlQuery::exec: database not open
QtTinySA v0.8.0
Found TinySA on /dev/ttyACM0
serial port exception
```

The GUI then starts up fine, but there is obviously no communication with the device. I have the `python3-pyqt5.qtsql` package installed, so the QSQLITE driver should be present on the system. There is, however, nothing to suggest that the "Database not open" message is in any way related to the "serial port exception". Can anyone suggest a way to diagnose this further?

Many thanks in advance!

**Comment by g4ixt** [comment]
I haven't seen that before but I wonder if it's because you have an older version of python as well as python3.
  
Try:  
  
python3 QtTinySA.py  
  
Do you have the correct serial module?  It should be 'python3-serial' from the repository.  Check for that first and if that fixes the serial connection we can look at the database issue.    The database connection is separate from the serial connection to the tinySA.

**Comment by jstarek** [comment]
Hi, first of all sorry for my late reply... I only have a single version of Python installed on the machine (which is 3.10.12), and I'm using Debian's / Ubuntu's package `python-is-python3` to provide the alias name of `python` for the invocation of `python3`. 

When testing with `python3 QtTinySA.py", I get the exact same output as stated above.

The `python3-serial` module I have installed is from Ubuntu's repo, and it's 3.5-1. 

The installation of the OS is new, and QtTinySA is the first project on this installation to use Python (or any serial port driver, for that matter). So I am pretty confident that there are no preexisting artifacts that complicate things.

**Comment by g4ixt** [comment]
If you try it without the TinySA connected, do you get anything like this?  
  
(base) ian@Study:~/Documents/GitHub/QtTinySA$ python3 QtTinySA.py
Database open: True
QtTinySA v0.9.0
TinySA not found

**Comment by arsinisa** [comment]
Just got on my hands a new TinySA Ultra and faced this same issue.
A workaround is to give permissions to /dev/ttyACM0 like this: ```sudo chmod a+rw /dev/ttyACM0```
But that works only as long as USB is not disconnected.

**Comment by jstarek** [comment]
Phew, now I feel stupid... @arsinisa 's hint to look at the permissions of the virtual serial port got me to resolve the first problem. However, @arsinisa , I would suggest to rather add the user you are using to the `dialout` group, who is typically the owner of that device and will already have rw permissions on it:

    $ sudo ls -l /dev/ttyACM0
    crw-rw---- 1 root dialout 166, 0 Dez 20 10:38 /dev/ttyACM0

@g4ixt, may I suggest adding a short hint like "make sure your user account has permission to access the /dev/ttyACM0 interface" into the prerequisites section of the README?

**Reply by g4ixt** [reply]
Yes I will add that.  But I thought it was the 'plugdev' group you need for USB not 'dialout', although when I type sudo ls -l /dev/ttyACM0 I get the same result as you.  If I type 'groups' I get this:

(base) ian@Study:~$ groups
ian adm cdrom sudo dip plugdev lpadmin lxd sambashare

**Comment by jstarek** [comment]
@g4ixt , my remaining issue was that database error, which I now was able to resolve. With the serial connection working, I focused on the error message `QSqlDatabase: available drivers: ` (which looks like an empty list) and found out that I had to install libqt5sql5-sqlite, which provides the actual, database-specific driver for the QTSQL package. This got everything working.

So perhaps that driver should also be included in the README under "dependencies" -- for many people, it will get auto-installed as a suggestion from another package, but that does not always seem to be the case.

Anyway, problem solved! :-) Thanks for your quick help here, and above all for the very nice software!
